### PR TITLE
add custom skin

### PR DESCRIPTION
**Pull request by david082321**

### CODE REVIEW

File: app/src/main/java/me/iacn/biliroaming/SettingDialog.kt [hunk 1]
```diff
@@ -278,6 +280,20 @@ class SettingDialog(context: Context) : AlertDialog.Builder(context) {
                         Log.toast("${e.message}")
                     }
                 }
+                SKIN_IMPORT -> {
+                    val file = File(currentContext.filesDir, "../files/skin.json")
```
Comment (yujincheng08) [inline app/src/main/java/me/iacn/biliroaming/SettingDialog.kt:284]:
```suggestion
                    val file = File(currentContext.filesDir, "skin.json")
```

File: app/src/main/java/me/iacn/biliroaming/SettingDialog.kt [hunk 2]
```diff
@@ -278,6 +280,20 @@ class SettingDialog(context: Context) : AlertDialog.Builder(context) {
...
+                    val uri = data?.data
+                    if (resultCode == RESULT_CANCELED || uri == null) return
+                    try {
+                        file.outputStream().use { out ->
+                            activity.contentResolver.openInputStream(uri)?.copyTo(out)
+                        }
+                    } catch (e: Exception) {
+                        e.printStackTrace()
```
Comment (yujincheng08) [inline app/src/main/java/me/iacn/biliroaming/SettingDialog.kt:292]:
```suggestion
                        Log.e(e)
```

File: app/src/main/java/me/iacn/biliroaming/hook/SkinHook.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import java.io.File
+import java.lang.reflect.Array as RArray
+
+/**
+ * Created by david on 2022/2/1
+ * Email ？
+ */
+
+class SkinHook (classLoader: ClassLoader) : BaseHook(classLoader) {
+    // （这里开始）直接抄 BangumiSeasonHook 的内容
```
Comment (yujincheng08) [inline app/src/main/java/me/iacn/biliroaming/hook/SkinHook.kt:15]: 放到 utils，不要抄

File: app/src/main/java/me/iacn/biliroaming/hook/SkinHook.kt [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                            // 从 导入 获取
+                            skin = File(currentContext.filesDir, "../files/skin.json")
+                                .readText().toJSONObject().putOpt("package_md5","")
+                                .toString().replace(
+                                    """package_md5":""""", """package_md5":null"""
+                                )
+                        } else if (sPrefs.getBoolean("skin", false)) {
+                            // 从 填写 获取
+                            skin = sPrefs.getString("skin_json", "")?.toJSONObject().toString()
+                        }
```
Comment (yujincheng08) [inline app/src/main/java/me/iacn/biliroaming/hook/SkinHook.kt:105]:
```suggestion
                        var skin = 
                        if (sPrefs.getBoolean("skin_import", false)) {
                            // 从 导入 获取
                            File(currentContext.filesDir, "../files/skin.json")
                                .readText().toJSONObject().putOpt("package_md5","")
                                .toString().replace(
                                    """package_md5":""""", """package_md5":null"""
                                )
                        } else if (sPrefs.getBoolean("skin", false)) {
                            // 从 填写 获取
                            sPrefs.getString("skin_json", "")?.toJSONObject().toString()
                        }
```

**Review by yujincheng08 — CHANGES_REQUESTED** [review]

File: app/src/main/java/me/iacn/biliroaming/hook/SkinHook.kt (hunk 3 above)
Comment (david082321) [reply app/src/main/java/me/iacn/biliroaming/hook/SkinHook.kt:15]: > 放到 utils，不要抄

放进去，然后 
`private val isSerializable by lazy {
    "com.bilibili.bangumi.data.page.detail.entity.BangumiUniformSeason\$\$serializer".findClassOrNull(
        mClassLoader
    ) != null
}`

的
`mClassLoader`
飘红

Comment (david082321) [reply app/src/main/java/me/iacn/biliroaming/hook/SkinHook.kt:15]: `private val serializerFeatures = lazy {
    val serializerFeatureClass =
        "com.alibaba.fastjson.serializer.SerializerFeature".findClassOrNull(mClassLoader)
（略）
}`
的
`mClassLoader`
也是

File: app/src/main/java/me/iacn/biliroaming/hook/SkinHook.kt [hunk 5]
```diff
@@ -91,17 +91,17 @@ class SkinHook (classLoader: ClassLoader) : BaseHook(classLoader) {
                         if (instance.generalResponseClass?.isInstance(body) == true) "data" else instance.responseDataField().value
                     val resultClass = body.getObjectField(dataField)?.javaClass
                     try {
-                        var skin = ""
+                        var skin =
```
Comment (yujincheng08) [inline app/src/main/java/me/iacn/biliroaming/hook/SkinHook.kt:94]: 格式化一下

Comment (david082321) [reply app/src/main/java/me/iacn/biliroaming/hook/SkinHook.kt]: 好了

**Comment by david082321** [comment]
# 主题格式范例

* 范例1
`{"id":114514,"name":"OvO","preview":"http://example.com/example.png","ver":1600000000,"package_url":"http://example.com/example.zip","data":{"color_mode":"light","color":"#ffffff","color_second_page":"#666666","side_bg_color":"#ababab","tail_color":"#e4e4e4","tail_color_selected":"#ffffff","tail_icon_ani":true,"tail_icon_ani_mode":"once","head_myself_mp4_play":"","tail_icon_mode":"img"}}`

* 范例2
`{"id":1919,"name":"","preview":"","ver":123456,"package_url":"https://biligo.org/example.zip","data":{"color_mode":"dark","color":"#ffffff","color_second_page":"#797979","side_bg_color":"#7a7a7a","tail_color":"#999999","tail_color_selected":"#222222","tail_icon_ani":true,"tail_icon_ani_mode":"once","head_myself_mp4_play":"","pub_btn_shade_color_top":"#797979","pub_btn_shade_color_bottom":"#666666","pub_btn_plus_color":"#ffffff","tail_icon_mode":"color","tail_icon_color":"#646569","tail_icon_color_selected":"#222222","tail_icon_color_dark":"#a4a5a7","tail_icon_color_selected_dark":"#f5f5f5"}}`

# 正文参数

| 参数名      	| 类型 	| 必要性 	| 内容                	| 填写内容范例                                                   	|
|-------------	|------	|--------	|---------------------	|----------------------------------------------------------------	|
| id          	| int  	| 必要   	| 编号                	|                                                                	|
| name        	| str  	| 必要   	| 显示名称            	| 随便写，反正不会用这个显示                                     	|
| preview     	| str  	| 必要   	| 预览图网址          	| 随便写，反正不会用这个显示<br>"http://example.com/example.png" 	|
| ver         	| int  	| 必要   	| 版本号              	|                                                                	|
| package_url 	| str  	| 必要   	| 图片的zip压缩包网址 	| 必须要写正确的压缩包地址<br>"http://example.com/example.zip"   	|
| data        	| obj  	| 必要   	| 下方有详细说明      	|                                                                	|

----

# ZIP内文件

## 图片

`head_bg`, `head_tab_bg`, `side_bg`, `side_bg_bottom`, `head_myself_bg`, `head_myself_squared_bg`, `tail_bg`, `tail_icon_main`, `tail_icon_channel`, `tail_icon_dynamic`, `tail_icon_shop`, `tail_icon_myself`, `tail_icon_pub_btn_bg`, `tail_icon_selected_main`, `tail_icon_selected_channel`, `tail_icon_selected_dynamic`, `tail_icon_selected_shop`, `tail_icon_selected_myself`, `tail_icon_selected_pub_btn_bg`

* 后缀自己加，可用 jpg 或 png，其他未测试
例如：
`tail_icon_myself.jpg`
`side_bg.png`

* 必要性：`head_bg` 及 `head_tab_bg` 为必要，其他非必要

## 视频

`head_myself_mp4_bg.mp4`

* 必要性：非必要

----

# `data` 参数

* 本节内容关于必要性的栏位，并未完整测试，可能有误

| 参数名                        	| 类型 	| 必要性   	|  内容                                                                       	| 填写内容范例                        	|
|-------------------------------	|------	|----------	|-----------------------------------------------------------------------------	|-------------------------------------	|
| color_mode                    	| str  	| 必要     	| 浅色/深色主题                                                               	| "light"<br>"dark"                   	|
| color                         	| str  	| 必要     	| 颜色<br>1. 使用十六进制颜色值<br>2. 颜色值的大小写不限，下同                	| "#ffffff"<br>"#98e233"<br>"#fFfFFf" 	|
| color_second_page             	| str  	| 必要     	| 颜色                                                                        	|                                     	|
| tail_color                    	| str  	| 必要     	| 颜色：底栏                                                                  	|                                     	|
| tail_color_selected           	| str  	| 必要     	| 颜色：底栏(选择时)                                                          	|                                     	|
| tail_icon_ani                 	| bool 	| 必要     	| 底栏图标动画                                                                	| true<br>false                       	|
| tail_icon_ani_mode            	| str  	| 必要     	| 底栏图标动画循环                                                            	| "once"                              	|
| head_myself_mp4_play          	| str  	| 必要     	| 视频循环                                                                    	| "once"<br>"loop"<br>""              	|
| tail_icon_mode                	| str  	| 必要     	| 底栏改图标(img)或改颜色(color)                                              	| "img"<br>"color"                    	|
| pub_btn_plus_color            	| str  	| 非必要   	| 颜色                                                                        	|                                     	|
| pub_btn_shade_color_bottom    	| str  	| 非必要   	| 颜色                                                                        	|                                     	|
| pub_btn_shade_color_top       	| str  	| 非必要   	| 颜色                                                                        	|                                     	|
| side_bg_color                 	| str  	| 非必要   	| 颜色：侧边栏                                                                	|                                     	|
| side_line_color               	| str  	| 非必要   	| 颜色                                                                        	|                                     	|
| tail_icon_color               	| str  	| 可能必要 	| 颜色：底栏图标颜色<br>若 tail_icon_mode = "color"，则必要                   	|                                     	|
| tail_icon_color_dark          	| str  	| 可能必要 	| 颜色：底栏图标颜色(夜间模式)<br>若 tail_icon_mode = "color"，则必要         	|                                     	|
| tail_icon_color_selected      	| str  	| 可能必要 	| 颜色：底栏图标颜色(选择时)<br>若 tail_icon_mode = "color"，则必要           	|                                     	|
| tail_icon_color_selected_dark 	| str  	| 可能必要 	| 颜色：底栏图标颜色(选择时)(夜间模式)<br>若 tail_icon_mode = "color"，则必要 	|                                     	|

**Comment by yujincheng08** [comment]
这个我还得再改改才能 merge

**Pull request closed without merging**